### PR TITLE
Integrate adaptive d

### PR DESCRIPTION
**Pull request by Alan-Penkar**
Added Adaptive D as a play.  Angles are still a bit questionable, but it runs as a play.  Let's try it!

### CODE REVIEW

File: soccer/gameplay/plays/offense/basic_122.py [hunk 1]
```diff
@@ -51,7 +51,18 @@ def __init__(self):
 
     @classmethod
     def score(cls):
-        return 10 if main.game_state().is_playing() else float("inf")
+        if (not main.game_state().is_playing()):
+            return float("inf")
+        if (not main.game_state().is_playing()):
```
Comment (jgkamat) [inline soccer/gameplay/plays/offense/basic_122.py]: I think this is a duplicate of the previous line

**Comment by kylestach** [comment]
@Jason27chan, would it be reasonable to add this to M1 for the semester, to be run in the test day a month from now? Ideally we can get this cleaned up and merged in as soon as possible to free you up for new plays work once #1314 is through.

File: soccer/gameplay/evaluation/ball.py [hunk 2]
```diff
@@ -4,6 +4,16 @@
 import math
 
 
+def we_are_closer():
```
Comment (Jason27chan) [inline soccer/gameplay/evaluation/ball.py]: @liquidmetal9015 have you written these functions already

File: soccer/gameplay/evaluation/linear_classification.py [hunk 3]
```diff
@@ -0,0 +1,34 @@
...
+
+# Classifies the object into two distinct class based on a cutoff value
+# Anything less than the cutoff is of class false, greater than the cutoff is of class true
+# 
+# @param input The vector of input features
+# @param weights The vector of weights to apply to the input features
+# @param bias The bias of the features-weight system
+# @param cutoff The number which splits the output score of the object into two classes
+# @param Returns tuple of the class (true or false) and the given score
+def binary_classification(input, weights, bias, cutoff):
```
Comment (Jason27chan) [inline soccer/gameplay/evaluation/linear_classification.py:21]: @liquidmetal9015 Have you written these functions already

File: soccer/gameplay/evaluation/ball.py (hunk 2 above)
Comment (Jason27chan) [reply soccer/gameplay/evaluation/ball.py]: I hope not because theyre getting removed lmao

File: soccer/gameplay/constants.py [hunk 4]
```diff
@@ -4,7 +4,7 @@
 DegreesToRadians = math.pi / 180.0
 RadiansToDegrees = 180.0 / math.pi
 OurChipRoll = 3  # Average distance at which a ball is slow enough to capture
-TheirChipping = (.1, .8)
```
Comment (Jason27chan) [inline soccer/gameplay/constants.py]: Why did you change this?? Refactor for consistency

File: soccer/gameplay/evaluation/ball.py [hunk 5]
```diff
@@ -65,6 +64,23 @@ def opponent_with_ball():
             return None
 
 
+# returns a Robot or None indicating which of us has the ball
```
Comment (Jason27chan) [inline soccer/gameplay/evaluation/ball.py]: Comment code in doxygen style http://www.doxygen.nl/manual/docblocks.html#pythonblocks

File: soccer/gameplay/evaluation/defensive_positioning.py [hunk 6]
```diff
@@ -242,3 +242,67 @@ def find_defense_positions(ignore_robots=[]):
     area_def_pos = create_area_defense_zones(ignore_robots)
 
     return area_def_pos, sorted_bot[0], sorted_bot[1]
+
+
+def goalside_mark_segment(mark_pos, robot, ball=False, kick_eval=None):
```
Comment (Jason27chan) [inline soccer/gameplay/evaluation/defensive_positioning.py:255]: Move comments to before method header like in http://www.doxygen.nl/manual/docblocks.html#pythonblocks

File: soccer/gameplay/evaluation/defensive_positioning.py [hunk 7]
```diff
@@ -242,3 +242,67 @@ def find_defense_positions(ignore_robots=[]):
...
+        intersections = sorted(tmp, key=lambda pt: pt.y, reverse=True)
+        
+        #Correction for when there is no segment because the opposing robot is inside a radius of our goalzone
+        if len(intersections)==0:
+            intersections.append(adjusted_mark_pos)
+
+        return robocup.Segment(adjusted_mark_pos, intersections[0]), shot_pt
+
+
+def ballside_mark_segment(mark_pos, ball_pos = None):
```
Comment (Jason27chan) [inline soccer/gameplay/evaluation/defensive_positioning.py]: Move documentation outside like in http://www.doxygen.nl/manual/docblocks.html#pythonblocks

File: soccer/gameplay/evaluation/linear_classification.py [hunk 8]
```diff
@@ -0,0 +1,34 @@
+
```
Comment (Jason27chan) [inline soccer/gameplay/evaluation/linear_classification.py]: Change commenting format to www.doxygen.nl/manual/docblocks.html#pythonblocks

File: soccer/gameplay/plays/defense/adaptive_defense.py [hunk 9]
```diff
@@ -0,0 +1,217 @@
...
+import evaluation.field
+import evaluation.linear_classification
+
+import tactics.positions.submissive_goalie as submissive_goalie
+import tactics.positions.wing_defender as wing_defender
+import tactics.wall as wall
+
+
+
+class AdaptiveDefense(standard_play.StandardPlay):
```
Comment (Jason27chan) [inline soccer/gameplay/plays/defense/adaptive_defense.py]: Add comments for class Change name to WingerWallDefense

File: soccer/gameplay/plays/defense/adaptive_defense.py [hunk 10]
```diff
@@ -0,0 +1,217 @@
...
+    # Weights / Bias for whether a opponent is a forward or winger
+    # Classifier returns true if it is a winger
+    WING_FORWARD_WEIGHTS = [-1.1, 1.8]
+    WING_FORWARD_BIAS    = 0
+    WING_FORWARD_CUTOFF  = 0
+
+    class State(Enum):
+        defending = 0
+
+    def __init__(self, defender_priorities=[20, 19, 18, 17, 16]):
```
Comment (Jason27chan) [inline soccer/gameplay/plays/defense/adaptive_defense.py]: magic numbers?

File: soccer/gameplay/plays/defense/adaptive_defense.py [hunk 11]
```diff
@@ -0,0 +1,217 @@
...
+        super().__init__(continuous=True)
+
+        goalie = submissive_goalie.SubmissiveGoalie()
+        goalie.shell_id = main.root_play().goalie_id
+        self.add_subbehavior(goalie, "goalie", required=True)
+
+        if len(defender_priorities) != 5:
```
Comment (Jason27chan) [inline soccer/gameplay/plays/defense/adaptive_defense.py]: ??

File: soccer/gameplay/plays/defense/adaptive_defense.py [hunk 12]
```diff
@@ -0,0 +1,217 @@
...
+            raise RuntimeError("defender_priorities must have a length of 5")
+
+        self.add_state(AdaptiveDefense.State.defending,
+                       behavior.Behavior.State.running)
+
+        self.add_transition(behavior.Behavior.State.start,
+                            AdaptiveDefense.State.defending, lambda: True,
+                            "immediately")
+
+        # TODO add goalie
```
Comment (Jason27chan) [inline soccer/gameplay/plays/defense/adaptive_defense.py]: Isn't this already added about 10 lines up?

File: soccer/gameplay/plays/defense/adaptive_defense.py [hunk 13]
```diff
@@ -0,0 +1,217 @@
...
+
+        self.aggression = 1
+        self.debug = True
```
Comment (Jason27chan) [inline soccer/gameplay/plays/defense/adaptive_defense.py]: Do we want this variable?

File: soccer/gameplay/plays/defense/adaptive_defense.py [hunk 14]
```diff
@@ -0,0 +1,217 @@
+import composite_behavior
```
Comment (Jason27chan) [inline soccer/gameplay/plays/defense/adaptive_defense.py]: Comment all the functions that arent related to the state machine

File: soccer/gameplay/plays/defense/adaptive_defense.py [hunk 15]
```diff
@@ -0,0 +1,217 @@
...
+
+        # TODO add goalie
+
+        self.aggression = 1
+        self.debug = True
+        self.kick_eval = robocup.KickEvaluator(main.system_state())
+        self.num_of_defenders = 5 # TODO: Make variable
+        self.wingers = []
+        self.forwards = []
+        self.max_wingers=3
```
Comment (Jason27chan) [inline soccer/gameplay/plays/defense/adaptive_defense.py]: move to static variable

File: soccer/gameplay/plays/testing/test_wing_defender.py [hunk 16]
```diff
@@ -0,0 +1,28 @@
+import main
+import play
+import behavior
+import constants
+import robocup
+
+import tactics.positions.wing_defender as wing_defender
+import evaluation.opponent as opponent_eval
+
+# Tests wing defender on the closest opponent robot to our goal
```
Comment (Jason27chan) [inline soccer/gameplay/plays/testing/test_wing_defender.py]: Doxygen format

File: soccer/gameplay/tactics/positions/wing_defender.py [hunk 17]
```diff
@@ -0,0 +1,113 @@
...
+import math
+import main
+import skills
+import enum
+import constants
+import role_assignment
+import evaluation.defensive_positioning
+from skills.move import Move
+
+class WingDefender(single_robot_behavior.SingleRobotBehavior):
```
Comment (Jason27chan) [inline soccer/gameplay/tactics/positions/wing_defender.py]: Docs

File: soccer/gameplay/tactics/positions/wing_defender.py [hunk 18]
```diff
@@ -0,0 +1,113 @@
...
+
+        self.add_state(WingDefender.State.defending, 
+                       behavior.Behavior.State.running)
+
+        self.add_transition(behavior.Behavior.State.start,
+                            WingDefender.State.defending, lambda: True,
+                            'immediately')
+        self.mv_pt = self.calc_move_point()
+        
+        # self.add_transition(
```
Comment (Jason27chan) [inline soccer/gameplay/tactics/positions/wing_defender.py:47]: Does this work? it would be cool if it does

File: soccer/gameplay/tactics/positions/wing_defender.py [hunk 19]
```diff
@@ -0,0 +1,113 @@
...
+        #   WingDefender.State.defending, WingDefender.State.intercepting,
+        #   lambda: self.ball_kicked,
+        #   'ball kicked, moving to intercept')
+
+    def execute_defending(self):
+        self._mark_pos = self._mark_robot.pos if self._mark_robot != None else self.mark_pos
+        self.mv_pt = self.calc_move_point()
+        self.robot.move_to(self.mv_pt)
+
+    def calc_move_point(self):
```
Comment (Jason27chan) [inline soccer/gameplay/tactics/positions/wing_defender.py:62]: docs

File: soccer/gameplay/tactics/positions/wing_defender.py [hunk 20]
```diff
@@ -0,0 +1,113 @@
...
+        self._goalside_ratio = value
+
+    @property
+    def distance(self):
+        return self._distance
+
+    @distance.setter
+    def distance(self, value):
+        self._distance = value
+
```
Comment (Jason27chan) [inline soccer/gameplay/tactics/positions/wing_defender.py]: extra line

File: soccer/gameplay/tactics/wall.py [hunk 21]
```diff
@@ -30,15 +38,37 @@ def __init__(self,
...
+                            "on shot")
+        self.add_transition(Wall.State.defense_wall,
+                            Wall.State.scramble, lambda:  evaluation.ball.we_are_closer() and evaluation.ball.moving_slow(),
+                            "ball free")
+        self.add_transition(Wall.State.scramble,
+                            Wall.State.defense_wall, lambda: not evaluation.ball.we_are_closer() or not evaluation.ball.moving_slow(),
+                            "ball captured")
+
+
+    def is_ball_not_free(self):
```
Comment (Jason27chan) [inline soccer/gameplay/tactics/wall.py]: Docs

File: soccer/gameplay/tactics/wall.py [hunk 22]
```diff
@@ -30,15 +38,37 @@ def __init__(self,
...
+        return main.ball().vel.mag() > 1 or min([(main.ball().pos - rob.pos).mag() for rob in main.system_state().their_robots]) <= min([(main.ball().pos - rob.pos).mag() for rob in main.system_state().our_robots])
+
+
+    def is_ball_shot(self):
```
Comment (Jason27chan) [inline soccer/gameplay/tactics/wall.py]: Docs

File: soccer/gameplay/tactics/wall.py [hunk 23]
```diff
@@ -47,10 +77,66 @@ def on_enter_defense_wall(self):
...
+    def execute_defense_wall(self):
+        if self.active_defenders < self.number_of_defenders:
+            #self._update_wall()
+            self._add_wall_defenders()
+            self.active_defenders = self.number_of_defenders
+        elif self.active_defenders > self.number_of_defenders:
+            self._remove_wall_defenders()
+            self.active_defenders = self.number_of_defenders
+
+    def _add_wall_defenders(self):
```
Comment (Jason27chan) [inline soccer/gameplay/tactics/wall.py:118]: docs

File: soccer/gameplay/tactics/wall.py [hunk 24]
```diff
@@ -47,10 +77,66 @@ def on_enter_defense_wall(self):
...
+            name="robot" + str(i)
+            pt = self.calculate_destination(i)
+            if i < self.active_defenders:
+                self.subbehavior_with_name(name).pos = pt
+            else:
+                self.add_subbehavior(
+                    skills.move.Move(pt),
+                    name=name)
+
+    def _remove_wall_defenders(self):
```
Comment (Jason27chan) [inline soccer/gameplay/tactics/wall.py:130]: docs

File: soccer/gameplay/tactics/wall.py [hunk 25]
```diff
@@ -47,10 +77,66 @@ def on_enter_defense_wall(self):
...
+        self.update_midpoint()
+        for i, priority in enumerate(self.defender_priorities[:self.active_defenders]):
+            name="robot" + str(i)
+            pt = self.calculate_destination(i)
+            if i < self.number_of_defenders:
+                self.subbehavior_with_name(name).pos = pt
+            else:
+                self.remove_subbehavior(name)
+
+    def _rebuild_wall(self):
```
Comment (Jason27chan) [inline soccer/gameplay/tactics/wall.py:142]: docs

File: soccer/gameplay/tactics/wall.py [hunk 26]
```diff
@@ -84,8 +175,12 @@ def mark_point(self):
     @mark_point.setter
     def mark_point(self, point):
         self._mark_point = point
+        self._update_wall()
+
+    def _update_wall(self):
```
Comment (Jason27chan) [inline soccer/gameplay/tactics/wall.py:193]: docs

File: soccer/gameplay/tactics/positions/wing_defender.py (hunk 18 above)
Comment (Jason27chan) [reply soccer/gameplay/tactics/positions/wing_defender.py:47]: TODO it

Comment (JNeiger) [reply soccer/gameplay/tactics/positions/wing_defender.py:47]: #1245

File: soccer/gameplay/evaluation/ball.py [hunk 27]
```diff
@@ -30,6 +30,16 @@ def is_in_our_goalie_zone():
...
 
+def we_are_closer():
+    return min([(main.ball().pos - rob.pos).mag() for rob in main.system_state().their_robots]) > min([(main.ball().pos - rob.pos).mag() for rob in main.system_state().our_robots])
+
+def opponent_is_much_closer():
+    return min([(main.ball().pos - rob.pos).mag() for rob in main.system_state().their_robots])*3 < min([(main.ball().pos - rob.pos).mag() for rob in main.system_state().our_robots])
+
+
+def moving_slow():
+    return main.ball().vel.mag() <= 1
```
Comment (kylestach) [inline soccer/gameplay/evaluation/ball.py]: This constant should definitely not be a magic number.

File: soccer/gameplay/evaluation/ball.py [hunk 28]
```diff
@@ -31,6 +30,16 @@ def is_in_our_goalie_zone():
     else:
         return False
 
+def we_are_closer():
+    return min([(main.ball().pos - rob.pos).mag() for rob in main.system_state().their_robots]) > min([(main.ball().pos - rob.pos).mag() for rob in main.system_state().our_robots])
+
+def opponent_is_much_closer():
+    return min([(main.ball().pos - rob.pos).mag() for rob in main.system_state().their_robots])*3 < min([(main.ball().pos - rob.pos).mag() for rob in main.system_state().our_robots])
```
Comment (kylestach) [inline soccer/gameplay/evaluation/ball.py]: Put a TODO here for when situational analysis exists.

**Comment by kylestach** [comment]
@Jason27chan I'll get that.

File: soccer/gameplay/evaluation/ball.py [hunk 29]
```diff
@@ -32,6 +31,28 @@ def is_in_our_goalie_zone():
         return False
```
Comment (HussainGynai) [inline soccer/gameplay/evaluation/ball.py:31]: This can be written as one return statement.
